Enable `_lazy_clone` between CPU and MPS

kurtamohler opened the pull request:
Adds `device` arg to `_lazy_clone` to enable lazy cloning data from one device to another. At the moment, only the following cases are supported:

* Source is a pinned CPU tensor and destination is MPS.
* Source is an MPS tensor and destination is CPU.
* Source and destination devices are the same.

This PR also adds support for pinned CPU tensors on MPS builds, which was not working properly before.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150569
* #150721
* __->__ #148408



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler commented on c10/core/impl/COW.cpp:
```diff
@@ -76,38 +84,69 @@ c10::intrusive_ptr<StorageImpl> lazy_clone_storage(StorageImpl& storage) {
+          c10::impl::cow::cow_deleter);
+      device_type = device.type();
+
+#ifdef USE_CUDA
```
Looks like `c10/core` is not supposed to depend on `c10/cuda`. `USE_CUDA` isn't even defined in c10. So I'll need to move any device-specific code like this out of here. I can do that by requiring that if the device is given to this function, then the allocator needs to be given as well. The caller to this function is in ATen, so we can get the allocator from that context

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler commented on aten/src/ATen/native/AutogradComposite.cpp:
```diff
@@ -91,14 +93,25 @@ bool _has_same_storage_numel(const at::Tensor& base, const at::Tensor& other) {
+Tensor _lazy_clone(Tensor const& self, optional<c10::Device> device_opt) {
+  optional<c10::Allocator*> allocator_opt = nullopt;
+  if (device_opt.has_value()) {
+    allocator_opt = at::empty({}, at::TensorOptions().device(device_opt.value())).storage().allocator();
```
This is not an ideal way to get the allocator, but there currently is no easy way to do it.

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

amjames commented on test/test_mps.py:
```diff
@@ -1493,6 +1493,122 @@ def test_exp(self, device="mps", dtype=torch.float):
+    # TODO: I would like to only have one copy of this test code and run it on
+    # both CUDA and MPS, but I'm not sure where is a good place to put it so
+    # that both test_torch and test_mps can have access to it. I guess
+    # probably somewhere in `torch.testing`
```
Since we don't have a test file for copy, and this is broader than MPS specifically. I would create a `test_lazy_clone.py` to hold this (and other tests related to lazy copy /COW behaviors). 

IMO adding a new file here would be better than burying the test somewhere that requires extra work to find it/remember where it lives.

kurtamohler replied:
Makes sense, I'll do that!

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler commented:
With the latest update, lazy cloning and materializing for between CPU and MPS now works. However, I have not yet added tests to read from a COW tensor whose data is on a different device than the target device for materialization, so that functionality may not necessarily work yet. I'll add a test for that and see what happens. Also, there are a few things I need to clean up in the code

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler commented:
I think I've reached a point where I'll need to understand more about MPS in order to progress.

I'd like to lay out an assumption I've made, in case I might be mistaken. Let's say we have the following:

```python
a = torch.tensor(..., device="device A")
b = a._lazy_clone(device="device B")
c = 2 * b
```

My assumption is that, since tensor `b` is semantically on device B, the multiply operation should be executed by device B, not device A. Furthermore, it should not cause a COW materialization, since multiply is just a read operation and `b` is not being written to.

I added a test that tries to perform read operations on a COW tensor whose target device differs from the original device, and it fails for both CPU-MPS and CPU-CUDA. (Of course, this is not a surprise for CPU-CUDA. It can't work because the GPU cannot access CPU address space, and vice versa.)

Presumably, there must be a way for MPS and CPU to read data that is stored in some shared memory location, and I guess that must be what this `getSharedBufferPtr` function is about: https://github.com/pytorch/pytorch/blob/2be97c7257a14e96bbe78e3b3050d16547d3a260/aten/src/ATen/mps/MPSAllocator.h#L296-L298

I will have to learn more about this in order to understand how to support read operations without materializing the COW data pointer. I'm wondering about a few specific questions:

* Can *any* MPS or CPU data be shared with the other device? Does the data buffer have to be initialized in some special way?
* Assuming that the CPU and MPS can both access the same data stored in a shared location (an assumption that, for all I know, might be false), why do we need COW anyway? Why couldn't we just have two tensors, one CPU and one MPS, reading and writing to the same data?
  * My guess at the answer is that we're using COW tensors as a way to avoid race conditions. If we had two different devices writing to the same data, we would need to use a mutex to avoid the race.

I'm wondering who I can ask about this stuff.

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler commented:
I think lazy cloning between CPU-MPS is essentially working now. I'll clean up the code and remove all the CPU-CUDA stuff (since I only added that in order to be able to run some of the code in this PR without a Mac). That shouldn't take too long, and then this PR will be ready for review.

One thing to point out: After this PR lands, there will still be a fair amount of work required to make sure that MPS tensors materialize in the correct situations. When I was implementing COW initially, I only ensured that it works correctly for CPU and CUDA, so the COW materialization tests in `test_ops.py` only run for those two devices. We will have to enable those tests for MPS as well and fix any data pointer const-correctness issues in the MPS code. That was a pretty large effort for CPU and CUDA, and I don't expect MPS to be any easier.

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

amjames commented on c10/core/impl/COW.cpp:
```diff
@@ -76,38 +108,95 @@ c10::intrusive_ptr<StorageImpl> lazy_clone_storage(StorageImpl& storage) {
+  c10::DeviceType device_type = storage.device_type();
+
+  if (device_opt.has_value()) {
+    allocator = allocator_opt.value();
```
If `allocator_opt` does not contain a value then this will throw. Should default to either the storage's allocator or the default allocator for the device?

amjames commented on c10/core/impl/COW.cpp:
```diff
@@ -133,12 +225,25 @@ C10_API void materialize_cow_storage(StorageImpl& storage) {
+          // NOTE: For MPS-to-CPU, `data_ptr.get()` gives an address in CPU
+          // space even though `src_device` is MPS. Likewise, for CPU-to-MPS,
+          // `data_ptr.get()` gives an address in MPS space even though the
+          // `src_device` is CPU. So both the src and dest pointers will always
+          // be in the address space of the dest device.
```
```suggestion
          // `src_device` is CPU. So both the data_ptr and new_data_ptr pointers will always
          // be in the address space of the dest device.
```

Helpful comment, but it's clearer if the naming is consistent with the surrounding code.

amjames commented on torch/csrc/Module.cpp:
```diff
@@ -2498,6 +2498,24 @@ Call this whenever a new thread is created in order to propagate values from
+      },
+      ("Gets the memory address of the Tensor's data pointer. If the device "
+       "is CPU and the allocator has unified memory on some other device, then "
+       "the CPU data address is resolved to the address spce of the other "
```
```suggestion
       "the CPU data address is resolved to the address space of the other "
```

amjames reviewed (changes requested):
IMO we should not land this until the lazy clones are materializing correctly on MPS. Otherwise it could put nightly users in a very strange state.

kurtamohler commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -3929,9 +3929,16 @@ class NativeCachingAllocator : public CUDAAllocator {
     C10_CUDA_CHECK(
         cudaMemcpy(dest, src, count, cudaMemcpyKind::cudaMemcpyDeviceToDevice));
+    if (sync) {
+      c10::cuda::device_synchronize();
```
That's probably true. Although nothing is actually exercising this code right now, and we might never need it. So maybe we should just throw an error saying this isn't implemented yet

amjames replied:
Yes, either way would work given this path is off topic for this PR.

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler commented:
I've made the requested changes, let me know if there is anything else

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/AutogradComposite.cpp:
```diff
@@ -91,18 +93,72 @@ bool _has_same_storage_numel(const at::Tensor& base, const at::Tensor& other) {
+    c10::DeviceType src_device_type = src_device.type();
+    c10::DeviceType dst_device_type = dst_device.type();
+    TORCH_CHECK(
+      (src_device_type == dst_device_type)
```
Do you properly support different device index below? if not, there should be a check here

kurtamohler replied:
I think different device indices would work, but we don't have any tests that exercise it right now (CI doesn't have any machines with two different MPS device indices). So it's probably best to raise an error to be safe

kurtamohler commented on c10/core/Allocator.h:
```diff
@@ -173,8 +173,22 @@ struct C10_API Allocator {
+  virtual const void* get_cpu_ptr_from_device_ptr(const void* device_ptr) const;
+  virtual void* get_device_ptr_from_cpu_ptr(void* cpu_ptr) const;
+  virtual const void* get_device_ptr_from_cpu_ptr(const void* cpu_ptr) const;
+  virtual bool has_unified_memory() const;
```
When CPU and Metal operators access the same location in the shared memory space, they use different addresses. (I wonder if there's a way to make them use the same address space?) So when we lazy clone an MPS tensor to pinned CPU, we need a way to translate the MPS address into CPU address space and set the output DataPtr to that. Likewise, we need to translate in the other direction for pinned CPU to MPS lazy clone. These functions provide an API for that

kurtamohler commented on c10/core/Allocator.h:
```diff
@@ -173,8 +173,22 @@ struct C10_API Allocator {
   // Note that this explicitly ignores any context that may have been
   // attached to the input data.
   //
+  // If `sync=true` is given, then the device will synchronize after the clone
```
My thought what that we need to synchronize when doing an inter-device materialization, in order to force any queued writes to the underlying data to happen before we try to copy that data to a different device. But now that I think about it more, that doesn't seem right, since we're already doing a sync before `_lazy_clone` returns, and it's supposed to be illegal to modify the COW data.

I'll disable the sync at the time of materialization to see if CI still passes. If it does, then I'll just remove the `sync` arg

kurtamohler replied:
Ok it looks like the sync during materialization is indeed unnecessary, so I'll just remove the `sync` arg

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update

[ghstack-poisoned]

kurtamohler committed:
Update on "Enable `_lazy_clone` between CPU and MPS"

Adds `device` arg to `_lazy_clone` to enable lazy cloning data from one device to another. At the moment, only the following cases are supported:

* Source is a pinned CPU tensor and destination is MPS.
* Source is an MPS tensor and destination is CPU.
* Source and destination devices are the same.

This PR also adds support for pinned CPU tensors on MPS builds, which was not working properly before.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

kurtamohler committed:
Update on "Enable `_lazy_clone` between CPU and MPS"

Adds `device` arg to `_lazy_clone` to enable lazy cloning data from one device to another. At the moment, only the following cases are supported:

* Source is a pinned CPU tensor and destination is MPS.
* Source is an MPS tensor and destination is CPU.
* Source and destination devices are the same.

This PR also adds support for pinned CPU tensors on MPS builds, which was not working properly before.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

kurtamohler committed:
Update on "Enable `_lazy_clone` between CPU and MPS"

Adds `device` arg to `_lazy_clone` to enable lazy cloning data from one device to another. At the moment, only the following cases are supported:

* Source is a pinned CPU tensor and destination is MPS.
* Source is an MPS tensor and destination is CPU.
* Source and destination devices are the same.

This PR also adds support for pinned CPU tensors on MPS builds, which was not working properly before.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

kurtamohler committed:
Update on "Enable `_lazy_clone` between CPU and MPS"

Adds `device` arg to `_lazy_clone` to enable lazy cloning data from one device to another. At the moment, only the following cases are supported:

* Source is a pinned CPU tensor and destination is MPS.
* Source is an MPS tensor and destination is CPU.
* Source and destination devices are the same.

This PR also adds support for pinned CPU tensors on MPS builds, which was not working properly before.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

kurtamohler committed:
Update on "Enable `_lazy_clone` between CPU and MPS"

Adds `device` arg to `_lazy_clone` to enable lazy cloning data from one device to another. At the moment, only the following cases are supported:

* Source is a pinned CPU tensor and destination is MPS.
* Source is an MPS tensor and destination is CPU.
* Source and destination devices are the same.

This PR also adds support for pinned CPU tensors on MPS builds, which was not working properly before.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/AutogradComposite.cpp:
```diff
@@ -91,18 +93,85 @@ bool _has_same_storage_numel(const at::Tensor& base, const at::Tensor& other) {
+      c10::DeviceIndex src_index = at::empty({}, at::TensorOptions().device(src_device)).device().index();
+      c10::DeviceIndex dst_index = at::empty({}, at::TensorOptions().device(dst_device)).device().index();
+
+      // NOTE: This case might already work, but we do not have any tests for it
```
IIRC, for cuda, it will only work if you have p2p access enabled. For non-cuda, not sure.
We should definitely keep this check

albanD commented on test/test_lazy_clone.py:
```diff
@@ -0,0 +1,216 @@
+# Owner(s): ["module: tests"]
```
This is not the right owner haha
We should have a module: cow on gitub and put that as the owner

albanD commented on test/test_lazy_clone.py:
```diff
@@ -0,0 +1,216 @@
+
+class TestLazyCloneDeviceType(TestCase):
+    def skip_if_lt_two_devices(self, device_type):
+        if device_type == "cuda":
+            if torch.cuda.device_count() < 2:
+                self.skipTest("Only one CUDA device found")
+        elif device_type == "mps":
+            if torch.mps.device_count() < 2:
+                self.skipTest("Only one MPS device found")
+        else:
+            self.skipTest(f"Index not supported for device type {device_type}")
```
`torch.accelerator.device_count()` can be used for all of this !

albanD commented on test/test_lazy_clone.py:
```diff
@@ -0,0 +1,216 @@
+
+    # Test that COW a tensor with a different target device can be used in read
+    # operations.
+    @skipCUDAIf(True, "Does not work for CUDA")
```
You can only instantiate on mps below if you skip for all other devices.

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -451,7 +451,7 @@
   result: auto_linear
 
-- name: _lazy_clone(Tensor self) -> Tensor
+- name: _lazy_clone(Tensor self, *, Device? device=None) -> Tensor
```
You need to update the formula to move back to the original device!

kurtamohler commented on aten/src/ATen/mps/MPSAllocator.mm:
```diff
@@ -810,10 +848,35 @@ bool waitForEvents(c10::ArrayRef<const void*> buffers) const override {
+  }
+
+  bool has_unified_memory() const override {
+    return m_has_unified_memory;
```
As far as I know, it's supposed to be true for any of the m-series Macs. `m_has_unified_memory` is given the value of the `MTLDevice`'s [`hasUnifiedMemory`](https://developer.apple.com/documentation/metal/mtldevice/hasunifiedmemory) property

kurtamohler committed:
Update

[ghstack-poisoned]
